packages/upload-client/src/fetch-with-upload-progress.js

```diff
@@ -1,8 +1,160 @@
+  const { onUploadProgress, body } = options
+
+  const rsp = new Response(body)
+  // @ts-expect-error web streams from node and web have different types
+  const source = iterateBodyWithProgress(rsp.body, onUploadProgress)
+  const stream = iterableToStream(source)
+  return {
+    ...options,
+    body: stream,
+  }
+}
+
+// HTTP/1.1 and browsers
+
+/* c8 ignore start */
+/**
+ *
+ * @param {string} url
+ * @param {import('./types.js').FetchOptions} init
+ */
+const fetchXhr = (url, { onUploadProgress, ...init }) => {
+  if (onUploadProgress) {
+    return /** @type {Promise<Response>} */ (
+      new Promise((resolve, reject) => {
+        const xhr = new XMLHttpRequest()
+        xhr.open(init.method || 'GET', url, true)
+        xhr.upload.addEventListener('progress', (e) =>
+          onUploadProgress({
+            total: e.total,
+            loaded: e.loaded,
+            lengthComputable: e.lengthComputable,
+            url,
+          })
+        )
+        xhr.upload.addEventListener('loadend', (e) =>
+          onUploadProgress({
+            total: e.total,
+            loaded: e.loaded,
+            lengthComputable: e.lengthComputable,
+            url,
+          })
+        )
+        xhr.onreadystatechange = () => {
+          if (xhr.readyState === 4) {
+            // @ts-expect-error doesn't have to match Response 100%
+            resolve({
+              status: xhr.status,
+              statusText: xhr.statusText,
+              body: xhr.response,
+              ok: ((xhr.status / 100) | 0) == 2,
+            })
+          }
+        }
+        xhr.onerror = (err) => reject(err)
+        Object.entries(init.headers || {}).forEach(([key, value]) =>
+          xhr.setRequestHeader(key, value)
+        )
+        /**
+         * @type {XMLHttpRequestBodyInit}
+         */
+        // @ts-expect-error ReadableStream as body is not supported by XHR
+        const body = init.body
+        xhr.send(body)
+      })
+    )
+  } else {
+    return fetch(url, init)
+  }
+}
+
+/* c8 ignore stop */
+
+// Deno supports H1 streaming
+const isNode =
+  typeof process !== 'undefined' &&
+  process.versions.node &&
+  !process.versions.bun &&
+  !process.versions.deno
+const isBrowser = typeof globalThis.XMLHttpRequest !== 'undefined'
 
 /**
  * @type {import('./types.js').FetchWithUploadProgress}
  */
-export const fetchWithUploadProgress = (url, init) => {
-  return ipfsUtilsFetch.fetch(url, init)
+export const fetchWithUploadProgress = (url, init = {}) => {
+  // https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming/nextHopProtocol
+  /**
+   * @type {string | undefined}
+   */
+  const protocol =
+    // @ts-expect-error nextHopProtocol is missing from types but is widely available
+    performance.getEntriesByType('resource')[0]?.nextHopProtocol
+  const preH2 = protocol !== 'h2' && protocol !== 'h2c' && protocol !== 'h3'
+
+  /* c8 ignore next 3 */
+  if ((isBrowser || preH2) && !isNode) {
```

Comment:
The test for `isBrowser` is an existence check for XHR, which could be `false`, but presumably `preH2` could be `true` and we could also not be in Nodejs...

We should not drop in here if XHR is not available.

Reply:
This might get confusing so I'll list the conditions/gotchas here:

1. Node.js fetch aka undici by default doesn't directly support HTTP/2, but supports body streaming either way.

2. Browser fetch supports streaming only in HTTP/2 (AFAIK)

3. Deno fetch is the same

4. Bun doesn't support HTTP/2 yet

5. All browsers implement XHR, so it's a safe assumption

perhaps a better set of conditions could be handled, not sure which one to handle all these cases (given that for some platforms streaming is available in both. for some in none and for others only in one)

also, in order to test this more thoroughly, a test case with HTTP/2 server perhaps should be added, but I haven't written one yet (should be trivial enough)